File: DKImagePickerControllerDemo_OC/DKImagePickerControllerDemo_OC/Source/DKImageManger/DKImageManager.m
//
//  DKImageManager.m
//  DKImagePickerControllerDemo_OC
//
//  Created by 赵铭 on 2017/6/23.
//  Copyright © 2017年 zm. All rights reserved.
//

#import "DKImageManager.h"
#import "DKGroupDataManager.h"
#import "DKAsset.h"

@interface DKImageManager()

@property (nonatomic, strong) PHCachingImageManager * manager;
@property (nonatomic, strong) PHImageRequestOptions * defaultImageRequestOptions;
@property (nonatomic, strong) PHVideoRequestOptions * defaultVideoRequestOptions;
@end


@implementation DKImageManager

- (id)init{
    self = [super init];
    if (self) {
        _manager = [[PHCachingImageManager alloc] init];
        _autoDownloadWhenAssetIsInCloud = YES;
    }
    return self;
}
- (DKGroupDataManager *)groupDataManager{
    if (!_groupDataManager) {
        _groupDataManager = [DKGroupDataManager new];
    }
    return  _groupDataManager;
}
- (PHImageRequestOptions *)defaultImageRequestOptions{
    if (_defaultImageRequestOptions == nil) {
        _defaultImageRequestOptions = [[PHImageRequestOptions alloc] init];
    }
    return _defaultImageRequestOptions;
}

- (PHVideoRequestOptions *)defaultVideoRequestOptions{
    if (_defaultVideoRequestOptions == nil) {
        _defaultVideoRequestOptions = [[PHVideoRequestOptions alloc] init];
        _defaultVideoRequestOptions.deliveryMode = PHVideoRequestOptionsDeliveryModeMediumQualityFormat;
    }
    return _defaultVideoRequestOptions;
}


+ (instancetype)shareInstance{
    static DKImageManager * shareInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        shareInstance = [[DKImageManager alloc] init];
    });
    return  shareInstance;
}

+ (void)checkPhotoPermissionWithHandle:(void(^)(BOOL granted))handle{
    if ([PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusAuthorized) {
        handle(YES);
    }else if ([PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusNotDetermined){
        [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
            dispatch_async(dispatch_get_main_queue(), ^{
                handle(status == PHAuthorizationStatusAuthorized);
            });
        }];
    }else{
        handle(NO);
    }
}


#pragma mark -- Fetch Image
- (void)fetchImageForAssetWith:(DKAsset *)asset
                          size:(CGSize)size
                 completeBlock:(void(^)(UIImage * image, NSDictionary * info))completeBlock

{
    
    [self fetchImageForAssetWith:asset size:size options:nil completeBlock:completeBlock];
}

- (void)fetchImageForAssetWith:(DKAsset *)asset
                          size:(CGSize)size
                   contentMode:(PHImageContentMode)contentMode
                 completeBlock:(void(^)(UIImage * image, NSDictionary * info))completeBlock{
    [self fetchImageForAssetWith:asset size:size options:nil contentMode:contentMode completeBlock:completeBlock];
}


- (void)fetchImageForAssetWith:(DKAsset *)asset
                          size:(CGSize)size
                       options:(PHImageRequestOptions *)options
                 completeBlock:(void(^)(UIImage * image, NSDictionary * info))completeBlock{
    [self fetchImageForAssetWith:asset size:size options:options contentMode:PHImageContentModeAspectFill completeBlock:completeBlock];
    
}



- (void)fetchImageForAssetWith:(DKAsset *)asset
                          size:(CGSize)size
                       options:(PHImageRequestOptions *)options
                   contentMode:(PHImageContentMode)contentMode
                 completeBlock:(void(^)(UIImage * image, NSDictionary * info))completeBlock

{
    if (!options) {
        options = self.defaultImageRequestOptions.copy;
    }
    [self.manager requestImageForAsset:asset.originalAsset targetSize:size contentMode:contentMode options:options resultHandler:^(UIImage * _Nullable result, NSDictionary * _Nullable info) {
        BOOL isInCloud = [info[PHImageResultIsInCloudKey] boolValue];
        if (isInCloud && !result && self.autoDownloadWhenAssetIsInCloud) {
            options.networkAccessAllowed = YES;
            [self fetchImageForAssetWith:asset size:size options:options contentMode:contentMode completeBlock:completeBlock];
        }else{
            completeBlock(result, info);
        }
    }];
    
}

- (void)fetchImageDataForAsset:(DKAsset *)asset
                       options:(PHImageRequestOptions *)options
                 completeBlock:(void(^)(NSData * imageData, NSDictionary * info))completeBlock{
    if (!options) {
        options = self.defaultImageRequestOptions.copy;
    }

    [self.manager requestImageDataForAsset:asset.originalAsset options:options resultHandler:^(NSData * _Nullable imageData, NSString * _Nullable dataUTI, UIImageOrientation orientation, NSDictionary * _Nullable info) {
        BOOL isInCloud = [info[PHImageResultIsInCloudKey] boolValue];
        if (isInCloud && !imageData && self.autoDownloadWhenAssetIsInCloud) {
            options.networkAccessAllowed = YES;
            [self fetchImageDataForAsset:asset options:options completeBlock:completeBlock];
        }else{
            completeBlock(imageData, info);
        }
        
    }];
}


- (void)fetchAVAsset:(DKAsset *)asset
       completeBlock:(void(^)(AVAsset * avAsset, NSDictionary * info))completeBlock{
}


- (void)fetchAVAsset:(DKAsset *)asset
             options:(PHVideoRequestOptions *)options
completeBlock:(void(^)(AVAsset * avAsset, NSDictionary * info))completeBlock
{
    if (!options) {
        options = self.defaultVideoRequestOptions.copy;
    }
    [self.manager requestAVAssetForVideo:asset.originalAsset options:options resultHandler:^(AVAsset * _Nullable avAsset, AVAudioMix * _Nullable audioMix, NSDictionary * _Nullable info) {
        BOOL isInCloud = [info[PHImageResultIsInCloudKey] boolValue];
        if (isInCloud && !asset && self.autoDownloadWhenAssetIsInCloud) {
            options.networkAccessAllowed = YES;
            [self fetchAVAsset:asset options:options completeBlock:completeBlock];
        }else{
            completeBlock(avAsset, info);
        }
    }];
}
- (void)stopCachingForAllAssets{
    [self.manager stopCachingImagesForAllAssets];
}
+ (CGSize)toPixel:(CGSize)size{
    CGFloat scale = [[UIScreen mainScreen] scale];
    return CGSizeMake(size.width * scale, size.height * scale);
}
- (void)invalidate{
    [self.groupDataManager invalidate];
    self.groupDataManager = nil;
}
@end

File: DKImageManager/DKImageManager.swift
//
//  DKImageManager.swift
//  DKImagePickerControllerDemo
//
//  Created by ZhangAo on 15/11/29.
//  Copyright © 2015年 ZhangAo. All rights reserved.
//

import Photos

open class DKBaseManager: NSObject {

	private let observers = NSHashTable<AnyObject>.weakObjects()
	
	open func addObserver(_ object: AnyObject) {
		self.observers.add(object)
	}
	
	open func removeObserver(_ object: AnyObject) {
		self.observers.remove(object)
	}
	
	open func notifyObserversWithSelector(_ selector: Selector, object: AnyObject?) {
		self.notifyObserversWithSelector(selector, object: object, objectTwo: nil)
	}
	
	open func notifyObserversWithSelector(_ selector: Selector, object: AnyObject?, objectTwo: AnyObject?) {
		if self.observers.count > 0 {
			DispatchQueue.main.async(execute: { () -> Void in
				for observer in self.observers.allObjects {
					if observer.responds(to: selector) {
						_ = observer.perform(selector, with: object, with: objectTwo)
					}
				}
			})
		}
	}

}

public func getImageManager() -> DKImageManager {
	return DKImageManager.sharedInstance
}

public class DKImageManager: DKBaseManager {
	
	public class func checkPhotoPermission(_ handler: @escaping (_ granted: Bool) -> Void) {
		func hasPhotoPermission() -> Bool {
			return PHPhotoLibrary.authorizationStatus() == .authorized
		}
		
		func needsToRequestPhotoPermission() -> Bool {
			return PHPhotoLibrary.authorizationStatus() == .notDetermined
		}
		
		hasPhotoPermission() ? handler(true) : (needsToRequestPhotoPermission() ?
			PHPhotoLibrary.requestAuthorization({ status in
				DispatchQueue.main.async(execute: { () in
					hasPhotoPermission() ? handler(true) : handler(false)
				})
			}) : handler(false))
	}
	
	static let sharedInstance = DKImageManager()
	
    private let manager = PHCachingImageManager()
	
	private lazy var defaultImageRequestOptions: PHImageRequestOptions = {
		let options = PHImageRequestOptions()
		
		return options
	}()
	
	private lazy var defaultVideoRequestOptions: PHVideoRequestOptions = {
		let options = PHVideoRequestOptions()
		options.deliveryMode = .mediumQualityFormat
		
		return options
	}()
	
	public var autoDownloadWhenAssetIsInCloud = true
	
    public lazy var groupDataManager: DKGroupDataManager! = {
        return DKGroupDataManager()
    }()
	
	public func invalidate() {
		self.groupDataManager.invalidate()
        self.groupDataManager = nil
	}
	
	public func fetchImageForAsset(_ asset: DKAsset, size: CGSize, completeBlock: @escaping (_ image: UIImage?, _ info: [AnyHashable: Any]?) -> Void) {
		self.fetchImageForAsset(asset, size: size, options: nil, completeBlock: completeBlock)
	}
	
	public func fetchImageForAsset(_ asset: DKAsset, size: CGSize, contentMode: PHImageContentMode, completeBlock: @escaping (_ image: UIImage?, _ info: [AnyHashable: Any]?) -> Void) {
			self.fetchImageForAsset(asset, size: size, options: nil, contentMode: contentMode, completeBlock: completeBlock)
	}

	public func fetchImageForAsset(_ asset: DKAsset, size: CGSize, options: PHImageRequestOptions?, completeBlock: @escaping (_ image: UIImage?, _ info: [AnyHashable: Any]?) -> Void) {
		self.fetchImageForAsset(asset, size: size, options: options, contentMode: .aspectFill, completeBlock: completeBlock)
	}
	
	public func fetchImageForAsset(_ asset: DKAsset, size: CGSize, options: PHImageRequestOptions?, contentMode: PHImageContentMode,
	                               completeBlock: @escaping (_ image: UIImage?, _ info: [AnyHashable: Any]?) -> Void) {
            let options = (options ?? self.defaultImageRequestOptions).copy() as! PHImageRequestOptions

            self.manager.requestImage(for: asset.originalAsset!,
                                      targetSize: size,
                                      contentMode: contentMode,
                                      options: options,
                                      resultHandler: { image, info in
                                        if let isInCloud = info?[PHImageResultIsInCloudKey] as AnyObject?
                                            , image == nil && isInCloud.boolValue && self.autoDownloadWhenAssetIsInCloud {
                                            options.isNetworkAccessAllowed = true
                                            self.fetchImageForAsset(asset, size: size, options: options, contentMode: contentMode, completeBlock: completeBlock)
                                        } else {
                                            completeBlock(image, info)
                                        }
            })
	}
	
	public func fetchImageDataForAsset(_ asset: DKAsset, options: PHImageRequestOptions?, completeBlock: @escaping (_ data: Data?, _ info: [AnyHashable: Any]?) -> Void) {
		self.manager.requestImageData(for: asset.originalAsset!,
		                                      options: options ?? self.defaultImageRequestOptions) { (data, dataUTI, orientation, info) in
												if let isInCloud = info?[PHImageResultIsInCloudKey] as AnyObject?
													, data == nil && isInCloud.boolValue && self.autoDownloadWhenAssetIsInCloud {
													let requestCloudOptions = (options ?? self.defaultImageRequestOptions).copy() as! PHImageRequestOptions
													requestCloudOptions.isNetworkAccessAllowed = true
													self.fetchImageDataForAsset(asset, options: requestCloudOptions, completeBlock: completeBlock)
												} else {
													completeBlock(data, info)
												}
		}
	}
	
	public func fetchAVAsset(_ asset: DKAsset, completeBlock: @escaping (_ avAsset: AVAsset?, _ info: [AnyHashable: Any]?) -> Void) {
		self.fetchAVAsset(asset, options: nil, completeBlock: completeBlock)
	}
	
	public func fetchAVAsset(_ asset: DKAsset, options: PHVideoRequestOptions?, completeBlock: @escaping (_ avAsset: AVAsset?, _ info: [AnyHashable: Any]?) -> Void) {
		self.manager.requestAVAsset(forVideo: asset.originalAsset!,
			options: options ?? self.defaultVideoRequestOptions) { avAsset, audioMix, info in
				if let isInCloud = info?[PHImageResultIsInCloudKey] as AnyObject?
					, avAsset == nil && isInCloud.boolValue && self.autoDownloadWhenAssetIsInCloud {
					let requestCloudOptions = (options ?? self.defaultVideoRequestOptions).copy() as! PHVideoRequestOptions
					requestCloudOptions.isNetworkAccessAllowed = true
					self.fetchAVAsset(asset, options: requestCloudOptions, completeBlock: completeBlock)
				} else {
					completeBlock(avAsset, info)
				}
		}
	}
    
    public func startCachingAssets(for assets: [PHAsset], targetSize: CGSize, contentMode: PHImageContentMode, options: PHImageRequestOptions?) {
        self.manager.startCachingImages(for: assets, targetSize: targetSize, contentMode: contentMode, options: options)
    }
    
    public func stopCachingAssets(for assets: [PHAsset], targetSize: CGSize, contentMode: PHImageContentMode, options: PHImageRequestOptions?) {
        self.manager.stopCachingImages(for: assets, targetSize: targetSize, contentMode: contentMode, options: options)
    }
    
    public func stopCachingForAllAssets() {
        self.manager.stopCachingImagesForAllAssets()
    }
}
